docs/configuration.rst
###################
Configuration Guide
###################

After installation, you will need to create a configuration file in a directory dedicated to ``pattoo``.

*********************************************
Setting the  Configuration Directory Location
*********************************************

Currently the configuration directory is automatically set when the installationscript is run


*********************
Configuration Options

*********************


There are two ways to configure ``pattoo``. These are the:

#. Quick Method
#. Expert Method

Quick Method
============

Use the quick method if you are new to ``pattoo``.

.. code-block:: bash
    $ sudo setup/pattoo_installation.py install configuration


The above command will set the most optimal defaults for your system for pattoo.

To guarantee success you will need to know the following beforehand.

#. ``db_name``: Database name
#. ``db_username``: Database username
#. ``db_password``: Database password
#. ``db_hostname``: Database hostname



Expert Method
=============

This section goes into configuration parameters in great detail.

Copy the Templates to Your Configuration Directory
--------------------------------------------------

Follow the steps in this section if you don't already have a valid configuration files in your ``PATTOO_CONFIGDIR`` directory.

Copy the template files in the ``examples/etc`` directory to the ``PATTOO_CONFIGDIR`` location.

**NOTE:** If a ``/path/to/configuration/directory/pattoo_server.yaml`` or ``/path/to/configuration/directory/pattoo.yaml`` file already exists in the directory then skip this step and edit the file according to the steps in following sections.

.. code-block:: bash

        $ cp examples/etc/pattoo_server.yaml.template \
            /path/to/configuration/directory/pattoo_server.yaml

        $ cp examples/etc/pattoo.yaml.template \
            /path/to/configuration/directory/pattoo.yaml

The next step is to edit the contents of both files.

Edit Your Configuration Files
-----------------------------

The ``pattoo`` server uses two configuration files:

#. ``pattoo.yaml``: Provides general configuration information for all ``pattoo`` related applications. ``pattoo.yaml`` also defines how ``pattoo`` agents should connect to the ``pattoo`` server APIs.
#. ``pattoo_server.yaml``: Provides configuration details for all the ``pattoo`` server's API daemons. These APIs accept data from ``pattoo`` agents and also provide data to ``pattoo`` related applications through your browser.

Take some time to read up on ``YAML`` formatted files if you are not familiar with them. A background knowledge is always helpful.

Server Configuration File
.........................

The ``pattoo_server.yaml`` file created from the template will have sections that you will need to edit with custom values. Don't worry, these sections are easily identifiable as they all start with ``PATTOO_``

**NOTE:** The indentations in the YAML configuration are important. Make sure indentations line up. Dashes '-' indicate one item in a list of items (if applicable).

.. code-block:: yaml

   pattoo_api_agentd:

       ip_bind_port: 20201
       ip_listen_address: 0.0.0.0

   pattoo_apid:

       ip_bind_port: 20202
       ip_listen_address: 0.0.0.0

   pattoo_ingesterd:

       ingester_interval: 3600
       batch_size: 500
       graceful_timeout: 10

   pattoo_db:
       db_pool_size: 10
       db_max_overflow: 10
       db_hostname: PATTOO_DB_HOSTNAME
       db_name: PATTOO_DB_NAME
       db_password: PATTOO_DB_PASSWORD
       db_username: PATTOO_DB_USERNAME

Server Configuration Explanation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This table outlines the purpose of each configuration parameter.

.. list-table::
   :header-rows: 1

   * - Section
     - Configuration Parameters
     - Description
   * - ``pattoo_api_agentd``
     -
     -
   * -
     - ``ip_listen_address``
     - IP address used by the ``pattoo_api_agentd`` daemon for accepting data from remote ``pattoo`` agents. Default of '0.0.0.0' which indicates listening on all available network interfaces. You can also use IPv6 nomenclature such as ``::``. The ``pattoo`` APIs don't support IPv6 and IPv4 at the same time.
   * -
     - ``ip_bind_port``
     - TCP port of used by the ``pattoo_api_agentd`` daemon for accepting data from remote ``pattoo`` agents. Default of 20201.
   * - ``pattoo_apid``
     -
     -
   * -
     - ``ip_listen_address``
     - IP address used by the ``pattoo_apid`` daemon for providing data to remote clients. Default of '0.0.0.0' which indicates listening on all available network interfaces. You can also use IPv6 nomenclature such as ``::``. The ``pattoo`` APIs don't support IPv6 and IPv4 at the same time.
   * -
     - ``ip_bind_port``
     - TCP port of used by the ``pattoo_apid`` daemon for providing data to remote clients. Default of 20202.
   * - ``pattoo_ingesterd``
     -
     -
   * -
     - ``ingester_interval``
     - The interval between checking for new agent files in the cache directory. Only valid if using the ``pattoo_ingesterd`` daemon.
   * -
     - ``batch_size``
     - The number of files to read per processing batch until all files are processed.
   * -
     - ``graceful_timeout``
     - The amount of time required for the ingester to finish processing data when the stop or restart command is excuted before it is forcefully stopped or restarted.
   * - ``pattoo_db``
     -
     -
   * -
     - ``db_hostname``
     - Hostname of the database server
   * -
     - ``db_username``
     - Username required for database access
   * -
     - ``db_password``
     - Password required for database access
   * -
     - ``db_name``
     - Name of database
   * -
     - ``db_pool_size``
     - This is the largest number of connections that will be keep persistently with the database
   * -
     - ``db_max_overflow``
     - Maximum overflow size. When the number of connections reaches the size set in ``db_pool_size``, additional connections will be returned up to this limit. This is the floating number of additional database connections to be made available.


Client Configuration File
.........................

The ``pattoo.yaml`` file created from the template will have sections that you will need to edit with custom values. Don't worry, these sections are easily identifiable as they all start with ``PATTOO_``

**NOTE:** The indentations in the YAML configuration are important. Make sure indentations line up. Dashes '-' indicate one item in a list of items (if applicable).

.. code-block:: yaml

   pattoo:
       log_level: debug
       log_directory: PATTOO_LOG_DIRECTORY
       cache_directory: PATTOO_CACHE_DIRECTORY
       daemon_directory: PATTOO_DAEMON_DIRECTORY
       system_daemon_directory: PATTOO_SYSTEM_DAEMON_DIRECTORY


Client Configuration Explanation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This table outlines the purpose of each configuration parameter.

.. list-table::
   :header-rows: 1

   * - Section
     - Configuration Parameters
     - Description
   * - ``pattoo``
     -
     -
   * -
     - ``log_directory``
     - Path to logging directory. Make sure the username running the daemons have RW access to files there.
   * -
     - ``log_level``
     - Default level of logging. ``debug`` is best for troubleshooting.
   * -
     - ``cache_directory``
     - Directory that will temporarily store data data from agents prior to be added to the ``pattoo`` database.
   * -
     - ``daemon_directory``
     - Directory used to store daemon related data that needs to be maintained between reboots
   * -
     - ``system_daemon_directory``
     - Directory used to store daemon related data that should be deleted between reboots. This should only be configured if you are running ``pattoo`` daemons as ``systemd`` daemons. The ``systemd`` daemon installation procedure automatically adjusts this configuration. This parameter defaults to the ``daemon_directory`` value if it is not configured.


pattoo/configuration.py
#!/usr/bin/env python3
"""Pattoo classes that manage various configurations."""

# Python imports
import os
import datetime

# Import project libraries
from pattoo_shared.configuration import ServerConfig
from pattoo_shared.configuration import search
from pattoo.constants import (
    PATTOO_API_WEB_NAME, PATTOO_API_AGENT_NAME,
    PATTOO_INGESTERD_NAME)


class ConfigAPId(ServerConfig):
    """Class gathers all configuration information.

    Only processes the following YAML keys in the configuration file:

        The value of the PATTOO_API_WEB_NAME constant

    """

    def __init__(self):
        """Initialize the class.

        Args:
            None

        Returns:
            None

        """
        # Instantiate the Config parent
        ServerConfig.__init__(self)

    def db_name(self):
        """Get db_name.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = 'pattoo_db'
        sub_key = 'db_name'

        # Process configuration
        result = search(
            key, sub_key, self._server_yaml_configuration)

        # Get result
        return result

    def db_username(self):
        """Get db_username.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = 'pattoo_db'
        sub_key = 'db_username'

        # Process configuration
        result = search(
            key, sub_key, self._server_yaml_configuration)

        # Get result
        return result

    def db_password(self):
        """Get db_password.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = 'pattoo_db'
        sub_key = 'db_password'

        # Process configuration
        if 'PATTOO_TRAVIS' in os.environ:
            result = ''
        else:
            result = search(
                key, sub_key, self._server_yaml_configuration)

        # Get result
        return result

    def db_hostname(self):
        """Get db_hostname.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = 'pattoo_db'
        sub_key = 'db_hostname'

        # Process configuration
        result = search(
            key, sub_key, self._server_yaml_configuration)

        # Get result
        return result

    def db_pool_size(self):
        """Get db_pool_size.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = 'pattoo_db'
        sub_key = 'db_pool_size'
        intermediate = search(
            key, sub_key, self._server_yaml_configuration, die=False)

        # Set default
        if intermediate is None:
            result = 10
        else:
            result = int(intermediate)
        return result

    def db_max_overflow(self):
        """Get db_max_overflow.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = 'pattoo_db'
        sub_key = 'db_max_overflow'
        intermediate = search(
            key, sub_key, self._server_yaml_configuration, die=False)

        # Set default
        if intermediate is None:
            result = 10
        else:
            result = int(intermediate)
        return result

    def ip_listen_address(self):
        """Get ip_listen_address.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = PATTOO_API_WEB_NAME
        sub_key = 'ip_listen_address'
        result = search(
            key, sub_key, self._server_yaml_configuration, die=False)

        # Default to 0.0.0.0
        if result is None:
            result = '0.0.0.0'
        return result

    def ip_bind_port(self):
        """Get ip_bind_port.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = PATTOO_API_WEB_NAME
        sub_key = 'ip_bind_port'

        # Get result
        intermediate = search(
            key, sub_key, self._server_yaml_configuration, die=False)
        if intermediate is None:
            result = 20202
        else:
            result = int(intermediate)
        return result

    def jwt_secret_key(self):
        """Get jwt_secret_key.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = PATTOO_API_WEB_NAME
        sub_key = 'jwt_secret_key'

        # Process configuration
        result = search(key, sub_key, self._server_yaml_configuration)

        # Ensures that jwt_secret_key is set
        if (result is None) or result.strip(' ') == '':
            raise Exception('Plese set JWT SECRET KEY in config file')

        # Get result
        return result

    def acesss_token_exp(self):
        """Gets access_token_exp.

        Parsing config where:
        minutes: 'm', hours: 'h', days: 'D', weeks: 'W', months: 'M'

        Args:
            None

        Return:
           exp: acesss token expiration time using datetie.timedelta

        """
        # Initialize key variables
        key = PATTOO_API_WEB_NAME
        sub_key = 'acesss_token_exp'

        # Process configuration
        result = search(key, sub_key, self._server_yaml_configuration)

        # Setting timedelta for result
        # Sets a default value if acesss_token_exp not found
        exp = self.__exp(result)

        if exp is None:
            exp = datetime.timedelta(minutes=15)

        return exp

    def refresh_token_exp(self):
        """Gets refresh_token_exp.

        Parsing config where:
        minutes: 'm', hours: 'h', days: 'D', weeks: 'W', months: 'M'

        Args:
            None

        Return:
            exp: refresh token expiration time using datetie.timedelta

        """
        # Initialize key variables
        key = PATTOO_API_WEB_NAME
        sub_key = 'refresh_token_exp'

        # Process configuration
        result = search(key, sub_key, self._server_yaml_configuration)

        # Setting timedelta for result
        # Sets a default value if acesss_token_exp not found
        exp = self.__exp(result)

        if exp is None:
            exp = datetime.timedelta(minutes=15)

        return exp

    def __exp(self, time_string):
        """Parses string to create datetime.timedelta object

        Parsing config where:
        minutes: 'm', hours: 'h', days: 'D', weeks: 'W', months: 'M'

        Args:
            time_string: string to be parsed containing a time period

        Return:
            exp: datetime.timedelta object for a given expiration time

        """
        exp = None
        time_string = time_string.split('_')

        # Creating datetime.timedelta object
        if len(time_string) == 2:
            time_duration = int(time_string[0])
            time_stamp = time_string[-1]

            if time_stamp == 'm':
                exp = datetime.timedelta(minutes=time_duration)
            elif time_stamp == 'h':
                exp = datetime.timedelta(hours=time_duration)
            elif time_stamp == 'D':
                exp = datetime.timedelta(days=time_duration)
            elif time_stamp == 'W':
                exp = datetime.timedelta(weeks=time_duration)
            elif time_stamp == 'M':
                exp = datetime.timedelta(months=time_duration)

        return exp


class ConfigAgentAPId(ServerConfig):
    """Class gathers all configuration information.

    Only processes the following YAML keys in the configuration file:

        The value of the PATTOO_API_WEB_NAME constant

    """

    def __init__(self):
        """Initialize the class.

        Args:
            None

        Returns:
            None

        """
        # Instantiate the Config parent
        ServerConfig.__init__(self)

    def ip_listen_address(self):
        """Get ip_listen_address.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = PATTOO_API_AGENT_NAME
        sub_key = 'ip_listen_address'
        result = search(
            key, sub_key, self._server_yaml_configuration, die=False)

        # Default to 0.0.0.0
        if result is None:
            result = '0.0.0.0'
        return result

    def ip_bind_port(self):
        """Get ip_bind_port.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = PATTOO_API_AGENT_NAME
        sub_key = 'ip_bind_port'

        # Get result
        intermediate = search(
            key, sub_key, self._server_yaml_configuration, die=False)
        if intermediate is None:
            result = 20202
        else:
            result = int(intermediate)
        return result

    def api_email_address(self):
        """GET API email address from yaml file.

        Args:
            None

        Returns:
            email (str): Email address of API
        """
        # Initialize key variables
        key = PATTOO_API_AGENT_NAME
        sub_key = 'api_encryption_email'

        result = search(
            key, sub_key, self._server_yaml_configuration, die=False)
        if result is None:
            result = 'pattoo_api@example.org'
        return result


class ConfigIngester(ServerConfig):
    """Class gathers all configuration information.

    Only processes the following YAML keys in the configuration file:

        The value of the PATTOO_INGESTERD_NAME constant

    """

    def __init__(self):
        """Initialize the class.

        Args:
            None

        Returns:
            None

        """
        # Instantiate the Config parent
        ServerConfig.__init__(self)

    def ingester_interval(self):
        """Get ingester_interval.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = PATTOO_INGESTERD_NAME
        sub_key = 'ingester_interval'
        result = search(
            key, sub_key, self._server_yaml_configuration, die=False)

        # Make sure we have an integer
        try:
            result = max(15, abs(int(result)))
        except:
            result = None

        # Default to 1 hour
        if result is None:
            result = 3600
        return result

    def graceful_timeout(self):
        """Get graceful_timeout.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = PATTOO_INGESTERD_NAME
        sub_key = 'graceful_timeout'
        result = search(
            key, sub_key, self._server_yaml_configuration, die=False)

        # Make sure we have an integer
        try:
            result = int(result)
        except:
            result = None

        # Default to 1 hour
        if result is None:
            result = 10
        return result

    def multiprocessing(self):
        """Get multiprocessing.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = PATTOO_INGESTERD_NAME
        sub_key = 'multiprocessing'
        _result = search(
            key, sub_key, self._server_yaml_configuration, die=False)

        # Return
        if _result is None:
            result = True
        else:
            result = bool(_result)
        return result

    def batch_size(self):
        """Get batch_size.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key varibles
        default = 500

        # Get result
        key = PATTOO_INGESTERD_NAME
        sub_key = 'batch_size'
        _result = search(
            key, sub_key, self._server_yaml_configuration, die=False)

        # Return
        if _result is None:
            result = default
        else:
            try:
                result = int(_result)
            except:
                result = default
        return result
